JeOS: refactor snapper test

asmorodskyi opened the pull request:
This commit simplifies determination of created snapshot ID.
Also it extracts common logic into helper function

asmorodskyi commented:
first successful VR - http://autobot/tests/637 
@Vogtinator , @gmoro - what do you think about this changes ?

asmorodskyi commented:
VR :  http://autobot/tests/638 - SLE 15 SP1

cfconrad commented on tests/console/snapper_jeos_cli.pm:
```diff
@@ -15,50 +15,37 @@ use testapi;
+    $self->rollback_and_reboot($openqalatest);
     assert_script_run("ls -l $file");
     assert_script_run("rm -v $file");
-    assert_script_run("snapper rollback $openqainit_snapshot");
```
To make the logic equal, we need a 
```
$self->rollback_and_reboot($openqainit);
assert_script_run("! ls -l $file");
```

asmorodskyi replied:
this is part which I was trying to explain in a call , I explicitly drop this two lines. Idea behind it was that during `snapper create ` you could create different types of snapshots but according to snapper documentation to get other than "single"  snapshot you need to explicitly specify this. Which we are not doing so this rollback is just repeating same thing which we already checked before

cfconrad replied:
Where do we rollback from `test` to `init` with deleted file?

asmorodskyi replied:
I agreed to insert this two lines

cfconrad commented on tests/console/snapper_jeos_cli.pm:
```diff
@@ -15,50 +15,37 @@ use testapi;
+use version_utils qw(is_sle);
+
+sub rollback_and_reboot {
+
```
why this extra line?

asmorodskyi replied:
removed

cfconrad commented on tests/console/snapper_jeos_cli.pm:
```diff
@@ -15,50 +15,37 @@ use testapi;
 use utils;
```
Copyright could be updated

asmorodskyi replied:
done

mloviska commented:
> LGTM,
> I think only one reboot+check is missing.
> 
> In general I would fill the file with some content and check that too via checksum or so. (but maybe in a separate commit/PR)

And we do not leave the test from the initial snapshot.
There were 3 transitions:
1) snap_id 1 to 18 ( no file snapshot )
2) 18 to 20 ( file exists )
3) 20 to 22 ( no file in 22 )

We have dropped no. 3, instead we just remove the file.

mloviska commented:
> In general I would fill the file with some content and check that too via checksum or so. (but maybe in a separate commit/PR)

I think this should be highlighted, as of now we are checking if some inode record exists, but we want to go deeper and verify if data exists as well. :)

Very nice point @cfconrad!

asmorodskyi commented:
> > In general I would fill the file with some content and check that too via checksum or so. (but maybe in a separate commit/PR)
> 
> I think this should be highlighted, as of now we are checking if some inode record exists, but we want to go deeper and verify if data exists as well. :)
> 
> Very nice point @cfconrad!

I will add this as idea into https://progress.opensuse.org/issues/42353 . But I would say that it should in separate

asmorodskyi commented:
> > LGTM,
> > I think only one reboot+check is missing.
> > In general I would fill the file with some content and check that too via checksum or so. (but maybe in a separate commit/PR)
> 
> And we do not exit the test case from the initial snapshot.
> There were 3 transitions:
> 
>     1. snap_id 1 to 18 ( no file snapshot )
> 
>     2. 18 to 20 ( file exists )
> 
>     3. 20 to 22 ( no file in 22 )
> 
> 
> We have dropped no. 3, instead we just remove the file.

so you saying that because of my changes we are not leaving system in state which was before test execution ? good enough argument to return back two lines which I dropped . 
Will do this and re-run VRs again

mloviska commented:
> > > LGTM,
> > > I think only one reboot+check is missing.
> > > In general I would fill the file with some content and check that too via checksum or so. (but maybe in a separate commit/PR)
> > 
> > 
> > And we do not exit the test case from the initial snapshot.
> > There were 3 transitions:
> > ```
> > 1. snap_id 1 to 18 ( no file snapshot )
> > 
> > 2. 18 to 20 ( file exists )
> > 
> > 3. 20 to 22 ( no file in 22 )
> > ```
> > 
> > 
> > We have dropped no. 3, instead we just remove the file.
> 
> so you saying that because of my changes we are not leaving system in state which was before test execution ? good enough argument to return back two lines which I dropped .
> Will do this and re-run VRs again

And you remove the file instead of go back to last snapshot without the file. 
You basically achieve the same, but different way. And since it is snapper test, I would rather go for rollback, then just remove the file from some snapshot and continue with other test cases.

asmorodskyi commented:
> 
> And you remove the file instead of go back to last snapshot without the file.
> You basically achieve the same, but different way. And since it is snapper test, I would rather go for rollback, then just remove the file from some snapshot and continue with other test cases.


hm you actually help me to understand a hidden bug in this test , which **was** there before : 

1. test was removing file https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L53 
2. and than reverting to snapshot where this file should not exists and checking that file is not existing .
For me it does not make any sense even without what I told about check for type of snapshot . 
So I would remove this line  https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L53  but will keep rollback , or I missing something ? WDYT @mloviska  , @cfconrad ?

mloviska commented:
> > And you remove the file instead of go back to last snapshot without the file.
> > You basically achieve the same, but different way. And since it is snapper test, I would rather go for rollback, then just remove the file from some snapshot and continue with other test cases.
> 
> hm you actually help me to understand a hidden bug in this test , which **was** there before :
> 
>     1. test was removing file https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L53
> 
>     2. and than reverting to snapshot where this file should not exists and checking that file is not existing .
>        For me it does not make any sense even without what I told about check for type of snapshot .
>        So I would remove this line  https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L53  but will keep rollback , or I missing something ? WDYT @mloviska  , @cfconrad ?

https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L54 and it rollbacked to the first snapshot created by the test case.

asmorodskyi commented:
> > > And you remove the file instead of go back to last snapshot without the file.
> > > You basically achieve the same, but different way. And since it is snapper test, I would rather go for rollback, then just remove the file from some snapshot and continue with other test cases.
> > 
> > 
> > hm you actually help me to understand a hidden bug in this test , which **was** there before :
> > ```
> > 1. test was removing file https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L53
> > 
> > 2. and than reverting to snapshot where this file should not exists and checking that file is not existing .
> >    For me it does not make any sense even without what I told about check for type of snapshot .
> >    So I would remove this line  https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L53  but will keep rollback , or I missing something ? WDYT @mloviska  , @cfconrad ?
> > ```
> 
> https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L54 and it rollbacked to the first snapshot created by the test case.

yes and the first snapshot is the one which **not** have the file . So we deleting a file and reverting to snapshot which does not have it. 
Can you clearly state what we checking by https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L54 ? My problem with this line that you can remove it but following check https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/console/snapper_jeos_cli.pm#L61 will still pass

asmorodskyi commented:
new VRs with current version : 
http://autobot/tests/639
http://autobot/tests/640

cfconrad commented:
For me it is just one of the combinations which might be tested. And yes there are combinations missing. But as you are doing a refactor, I would not remove/add functionality.

And for me it makes sense to test these to
```
test->init => check !file
test->rm file->init => check !file
```
It's like
```
init->test => check file
init->rm file -> test => check file
```
But we just missing the second case.

asmorodskyi commented:
@cfconrad , @mloviska - so latest commited version actually have all need checks from your perspective ? Or it is still missing something ?

mloviska commented:
> new VRs with current version :
> http://autobot/tests/639
> http://autobot/tests/640

I cannot reach these runs :(.
Maybe gitref and let's use o.s.d?
